docker-storage-setup.sh

```diff
@@ -865,14 +969,102 @@ setup_storage() {
    Fatal "Storage is already configured with ${current_driver} driver. Can't configure it with ${STORAGE_DRIVER} driver. To override, remove $DOCKER_STORAGE and retry."
   fi
 
-  # Set up lvm thin pool LV
+  # If a user decides to setup both (a) and (b):
+  # a) lvm thin pool for devicemapper.
+  # b) a separate volume for docker root.
+  # (a) will have a higher priority than (b) and will be setup first.
+
+  # Set up lvm thin pool LV.
   if [ "$STORAGE_DRIVER" == "devicemapper" ]; then
     setup_lvm_thin_pool
   else
       write_storage_config_file
   fi
+
+  # If docker root is on a separate volume, setup that.
+  if ! setup_docker_root_lv_fs; then
+    echo "Failed to setup docker root volume."
+    return 1
+  fi
+}
+
+restore_selinux_context() {
+  local dir=$1
+
+  if ! restorecon -R -v $dir; then
+    echo "restorecon -R -v $dir failed."
+    return 1
+  fi
 }
 
+# Inherit the same selinux context as $DOCKER_ROOT_DIR (Default:/var/lib/docker).
+setup_selinux_context_root_dir() {
+  local dir=$1
+  local policy
+
+  [ "$dir" == "$DOCKER_ROOT_DIR" ] && return 0
+
+  # If there is already a labeling policy for the dir, don't try to add one.
+  policy=`semanage fcontext -l -C | grep "^$dir =" | wc -l`
+  if [ $? -ne 0 ];then
+    echo "Querying existing selinux rules for $dir failed."
+    return 1
+  fi
+
+  # If no policy exists, add one.
+  if [ "$policy" == "0" ]; then
+    if ! semanage fcontext -a -e $DOCKER_ROOT_DIR $dir; then
+      echo "semanage fcontext -a -e $DOCKER_ROOT_DIR $dir failed."
+      return 1
+    fi
+  fi
+}
+
+get_docker_root_dir(){
+    local flag=false
+    options=$(grep -e "^OPTIONS" /etc/sysconfig/docker|cut -d"'" -f 2)
+    for opt in $options
+    do
+        if [ "$flag" = true ];then
+           DOCKER_ROOT_DIR=$opt
+           flag=false
+           continue
+        fi
+        if [[ "$opt" == --graph* || "$opt" == -g* ]];then
+           if [[ "$opt" == *"="* ]];then
```

Comment:
This is not a valid option. I just set invalid option --graph-driver in /etc/sysconfig/docker and this logic still took it.